code/protsgix.c

```diff
@@ -71,47 +72,55 @@ static struct sigaction sigNext;
 
 static void sigHandle(int sig, siginfo_t *info, void *uap)  /* .sigh.args */
 {
-  int e;
-  /* sigset renamed to asigset due to clash with global on Darwin. */
-  sigset_t asigset, oldset;
-  struct sigaction sa;
-
-  AVER(sig == PROT_SIGNAL);
-
-  if(info->si_code == SEGV_ACCERR) {  /* .sigh.check */
-    AccessSet mode;
-    Addr base;
-    MutatorContextStruct context;
-
-    MutatorContextInitFault(&context, info, (ucontext_t *)uap);
-
-    mode = AccessREAD | AccessWRITE; /* .sigh.mode */
-
-    /* We assume that the access is for one word at the address. */
-    base = (Addr)info->si_addr;   /* .sigh.addr */
-
-    /* Offer each protection structure the opportunity to handle the */
-    /* exception.  If it succeeds, then allow the mutator to continue. */
-    if(ArenaAccess(base, mode, &context))
-      return;
-  }
-
-  /* The exception was not handled by any known protection structure, */
-  /* so throw it to the previously installed handler.  That handler won't */
-  /* get an accurate context (the MPS would fail if it were the second in */
-  /* line) but it's the best we can do. */
-
-  e = sigaction(PROT_SIGNAL, &sigNext, &sa);
-  AVER(e == 0);
-  sigemptyset(&asigset);
-  sigaddset(&asigset, PROT_SIGNAL);
-  e = sigprocmask(SIG_UNBLOCK, &asigset, &oldset);
-  AVER(e == 0);
-  kill(getpid(), PROT_SIGNAL);
-  e = sigprocmask(SIG_SETMASK, &oldset, NULL);
-  AVER(e == 0);
-  e = sigaction(PROT_SIGNAL, &sa, NULL);
-  AVER(e == 0);
+  ERRNO_SAVE {
+    int e;
+    /* sigset renamed to asigset due to clash with global on Darwin. */
+    sigset_t asigset, oldset;
+    struct sigaction sa;
+
+    AVER(sig == PROT_SIGNAL);
+
+    if (info->si_code == SEGV_ACCERR) {  /* .sigh.check */
+      AccessSet mode;
+      Addr base;
+      MutatorContextStruct context;
+
+      MutatorContextInitFault(&context, info, (ucontext_t *)uap);
+
+      mode = AccessREAD | AccessWRITE; /* .sigh.mode */
+
+      /* We assume that the access is for one word at the address. */
+      base = (Addr)info->si_addr;   /* .sigh.addr */
+
+      /* Offer each protection structure the opportunity to handle the */
+      /* exception.  If it succeeds, then allow the mutator to continue. */
+      if (ArenaAccess(base, mode, &context))
+        goto done;
+    }
+
+    /* The exception was not handled by any known protection structure, */
+    /* so throw it to the previously installed handler.  That handler won't */
+    /* get an accurate context (the MPS would fail if it were the second in */
+    /* line) but it's the best we can do. */
+
+    e = sigaction(PROT_SIGNAL, &sigNext, &sa);
+    AVER(e == 0);
+    e = sigemptyset(&asigset);
+    AVER(e == 0);
+    e = sigaddset(&asigset, PROT_SIGNAL);
+    AVER(e == 0);
+    e = sigprocmask(SIG_UNBLOCK, &asigset, &oldset);
```

Comment:
It can be `sigprocmask` not `pthread_sigmask` because this code runs even in the single-threaded case and because we don't care what thread the SEGV is delivered to.